### ISSUE TITLE
Add support for Persistent volumes updates

### ISSUE THREAD

**Issue by ppatierno** [strimzi/strimzi#152]
Support for handling the cluster configmap update and how it works on the PVCs side.

**Comment by ppatierno** [comment]
Thinking more about this feature, taking into account the current `storage` data field configuration (in JSON format) in the cluster ConfigMap, following my thoughts ...

The user could want to switch `type` from `ephemeral` to `persistent-claim`. It means that from the current "emptyDir", he wants to switch to use a persistent volume so that a PersistentVolumeClaim is needed. It's possible updating the StatefulSets with the addition of a VolumeClaimTemplate instead of the "emptyDir" volume and a rolling update is needed in this case, shutting down current pods and spinning up new ones but .... I was wondering if it's a real case that we should support and I'm not quite sure about that. Maybe someone would want to start with `ephemeral` for testing and then switching to `persistent-claim` for production without re-deploying the cluster (of course losing all the data in the previously "emptyDir") ?

The opposite case, so switching `type` from `persistent-claim` to `ephemeral`, sounds more weird to me. Who wants to start persisting messages and then moving to an environment where storage is ephemeral so useful just for testing ?

In case of `persistent-claim`, maybe it's possible that the user would to increase the volume `size`. It makes sense to me but this feature in Kubernetes seems to be still a proposal and it depends on the used file system (from the cloud provider for example). More info [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/grow-volume-size.md).

Finally, always in the `persistent-claim` case, changing `class` and `selector` means asking to Kubernetes to deploy a new volume (or searching for a new volume with the required selector filter) but even in this case no data is "automagically" copied from the current volume to the new one. In any case this type of change at runtime seems to be not supported by Kubernetes. It means updating the StatefulSets (so the VolumeClaimTemplate) and doing a rolling update.

The last operation could be switching `delete-claim` the flag (when using the `persistent-claim`). This change effects how the cluster controller behaves when ConfigMap deletion happens and if it has to delete or not PVCs. It could make sense if the user changes his mind on what to do with PVCs when deleting the cluster.

In conclusion ... does it make sense to support the update operation for persistent volumes ? maybe just the `delete-claim` flag change ?  Of course the user can just edit the ConfigMap but the cluster controller could be able to detect that it's a storage related change refusing it. I mean, logging some warning (and maybe sending Kubernetes Events) but re-updating the ConfigMap with the configuration in use.

@scholzj @tombentley any thoughts ?

**Comment by ppatierno** [comment]
After talking on the public official Strimzi [IRC](https://webchat.freenode.net/?randomnick=1&channels=strimzi&uio=d4) channel we agree that : 

- switching the `type` means having a complicated migration of the data from a volume to another and for the time being it's not a feature we need, as real use case. 
- the only reasonable change to support is the `delete-claim` flag switching if the user changes his mind on what to do with PVCs when deleting the cluster.
- all the other changes are rejected, logging a warning but leaving the current `storage` configuration.

**Comment by ppatierno** [comment]
Starting to develop this, it turned out that updating the `delete-claim` flag is not a matter of updating the related annotation in the PersistentVolumeClaim but we need to update the annotation in the VolumeClaimTemplate. Consider the following simple scenario : 

- we start with `delete-claim` = true
- annotation in VolumeClaimTemplate is true and so the annotation in the related PersistentVolumeClaim
- update on the cluster CM setting `delete-claim` = false
- currently the PersistentVolumeClaim is patched setting the annotation to false
- delete the cluster CM (what we expect is that the PVCs are not delete thanks to the previous update but instead they are deleted ...)

The PVCs deletion happens because the DeleteXXXX operation use the `fromStatefulSet` method for building the XXXCluster instance and in the previous scenario we didn't update the VolumeClaimTemplate but only the related PersistentVolumeClaim.

Patching the VolumeClaimTemplate is not possible : 

> Reason: StatefulSet.apps "my-cluster-kafka" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas' and 'containers' are forbidden.

at this point I think that the PVCs update should just be an operation to reject. @scholzj wdyt ?

**Comment by scholzj** [comment]
These _forbidden: updates to statefulset spec ..._ mesages are strange ... we already update much more than this :-)

I think you would be able to change the code to recover this setting not from the VolumeClaimTemplate but directly from PersistentVolumeClaim. But we would still need to update the template because of the situation when you change the `delete-claim` field and afterwards do scale up ... we need the new PVCs to have the right settings.

Would it help if this setting actually moved to the StatefulSet annotation? We anyway always start either with the StatefulSet or with ConfigMap. So maybe we can store it there and it will be easier to update.

**Comment by ppatierno** [comment]
> I think you would be able to change the code to recover this setting not from the VolumeClaimTemplate but directly from PersistentVolumeClaim

Yes it was my first idea but in any case the StatefulSets needed to be update to be consistent with the PVCs as you said.

> Would it help if this setting actually moved to the StatefulSet annotation? We anyway always start either with the StatefulSet or with ConfigMap. So maybe we can store it there and it will be easier to update.

Yes I think so

**Comment by ppatierno** [comment]
The forbidden updates drive to a problem.
The user updates the cluster CM, for example he sets `type` from `persistent-claim` to `ephemeral` but the cluster controller rejects this change without patching any resources but just logging a warning message.
The actual problem is that now the cluster CM is updated and not aligned with the real resources status in the cluster (i.e. cluster CM says `storage={ "type"="ephemeral" }` but the Kafka cluster is still working with `type`=`persistent-claim`).
What I have in mind is that the cluster controller has to patch the cluster CM with the real resources status (in this case storage only) in order to re-align the CM with the real deployment.

@scholzj any thoughts on that ?

**Comment by scholzj** [comment]
My view ... I don't think we should update the config maps. That fixes the problem of being out of sync but doesn't give the feedback to the user. We need the user to fix the config map. Its not an easy task but I think we should log the message, we should raise the event, we might be able to show it in some console later. But this is not the topic controller where we synchronise the operator actions between two sources. So I don't think we should touch the config map.

Maybe we can have a look how other operators deal with this. There are plenty of operators and they must have run into this kind of issue as well.

Maybe we should move forward with the idea which you started with the storage and have the configuration for the actual state stored in an annotation to have it somewhere in valid state. Not sure how exactly we would use it, but at least you can look based on which configuration was the cluster assembled.

**Comment by ppatierno** [comment]
The initial storage configuration with which the cluster was assembled is available through the Statefulsets and eventually the related VolumeClaimTemplate. What's the annotation you are talking about ? The only annotation used now is for the delete-claim in the StatefulSets as you suggested.

**Comment by scholzj** [comment]
Yeah ... we have already the `delete-claim` annotation. We can in theory put the whole config map into an annotation.

**Comment by ppatierno** [comment]
But when the cluster controller rejects the changes from the configmap, the "valid state" is still in the Statefulsets (delete-claim annotation) and the related VolumeClaimTemplate (size, storage class, selector) if "persistent". Sorry but I can't understand the reason for storing the entire storage configuration in an annotation :(

**Comment by scholzj** [comment]
Well, you would have it in some easily readable form ... even fairly readable for human. But as I said ... I'm not clear about the usage for that ... its not something for this issue for sure. It just a thought, idea.  

I think right now we should just log error since we do not have anything better. Firing an event as topic controller does should be also quite easy to do.

**Comment by ppatierno** [comment]
I have just opened a PR for this issue, do we want to take into account firing Kubernetes events now or opening an issue for the future ?

**Comment by scholzj** [comment]
I would say to leave it for separate issue - but I don't know how much effort it is (I assume its non-trivial effort).

**Comment by ppatierno** [comment]
I would leave it for a separate issue in order to try to use common code from topic-controller. It could be one of the starting points for sharing code. Otherwise we can work on the new issue in any case. I'm going to open it.

**Comment by scholzj** [comment]
Yeah, it could. Also when we work on that we can add that to some more places I guess, not only here.

**Comment by ppatierno** [comment]
Agree.

**Comment by ppatierno** [comment]
Closed by #176
